ctapipe/image/cleaning.py

```diff
@@ -2,16 +2,65 @@
 Image Cleaning Algorithms (identification of noisy pixels)
 """
 
-__all__ = ['tailcuts_clean', 'dilate']
+__all__ = ['tailcuts_clean', 'dilate', 'fact_image_cleaning', 'ImageCleaning']
 
 import numpy as np
 from scipy.sparse.csgraph import connected_components
-
-
-def tailcuts_clean(geom, image, picture_thresh=7, boundary_thresh=5,
-                   keep_isolated_pixels=False,
-                   min_number_picture_neighbors=0):
-
+from ctapipe.core import Component
+from traitlets import Dict, CaselessStrEnum
+
+
+class ImageCleaning(Component):
+    '''
+    A component to make image cleaning configurable using traitlets.
+    '''
+
+    key_type = CaselessStrEnum(
+        default_value='camera_id',
+        values=['camera_id', 'telescope_id'],
+        help='If the keys in settings are telescope ids or camera ids'
+    ).tag(config=True)
+    settings = Dict(
+        help='A dict mapping either camera_id or telescope_id to cleaning settings.'
+        ' This are keyword arguments for the actual cleaning function plus the'
+        ' the key algorithm which can be either "tailcuts" or "fact"'
+    ).tag(config=True)
+
+    def __call__(self, geom, image, arrival_times=None, telescope_id=None):
+
+        if self.key_type == 'camera_id':
+            kwargs = self.settings.get(geom.cam_id, {})
+        elif self.key_type == 'telescope_id':
+            if telescope_id is None:
+                raise ValueError(
+                    'telescope_id is required if using key_type="telescope_id"'
+                )
+            kwargs = self.settings.get(telescope_id, {})
+
+        algorithm = kwargs.pop('algorithm', 'tailcuts')
+
+        if algorithm == 'tailcuts':
```

Comment:
We adopt a factory approach for other algorithms, wouldn't it make sense to keep the same design across the pipeline? By having a Tailcut and FactImageCleaning subclass to the ImageCleaning baseclass? 

I can understand why you would do it like this (the functions exist, the classes don't, they are simple methods, and theres only two of them at the moment) but I'm just suggesting this from a consistency point of view.

Reply:
I am not sure I understand Jasons point here correctly or if I have another point. I would propose to refrain from inventing new names for things (even if they are shorter and nicer) and thus avoid mental mapping, i.e. when the user wants the function `fact_image_cleaning` to be called, they should in my opinion have to write this:

    settings={
        'LSTCam': {'algorithm': 'fact_image_cleaning', 'picture_threshold': 10, 'boundary_threshold': 5}
    }

and not the shorter form:

    settings={
        'LSTCam': {'algorithm': 'fact', 'picture_threshold': 10, 'boundary_threshold': 5}
    }

Another reason for using this longer form would be: Assume you are a young bachelor starting here, and somebody gives you an already working analysis with such a config file as a starting point for your work. If you are interested in understanding the cleaning, it would be pretty cool if you could just search the whole codebase for the name after "algorithm" and actually have a single hit which happens to be the exact function you are looking for...

Reply:
I agree with @dneise on his point.

My point was that this `if... else...` clause is very similar use case to the factory approach we use for other components like the `ChargeExtractor`. It would be good to stay consistent with design across ctapipe.

Reply:
I must say, I have been very inspired by a talk on KM3Pipe at PyGamma19 this week. Their pipeline is kept quite simple, and I like their configuration approach for their "modules". I include the link here for those who are curious: https://indico.cern.ch/event/783425/contributions/3328264/

Reply:
Yes, we met tamas at last years Asterics Workshop. 

We briefly talked about the possibility of just using km3pipe and writing cta modules.
I like this design a lot.

Reply:
The only think I don't like about the modular approach they use is the same as for the HESS code: it's hard to then run them in a more simple way (simple for-loop) and debug them, but a lot of it is very nice. Originally the concept for ctapipe (back when it was also mostly C++) was such a modular flow-based approach. We even had ctapipe-flow for that as well, but in the end it's sometimes a bit overkill, and we can wrap in such a system later.

Reply:
I agree with avoiding the modular approach. The things we will wish to do in our pipeline typically have a set order, and its more clear to define what is done at each stage or data level. The freedom we allow the user to set is exactly which algorithm implementation is used for each of the fixed stages (`ChargeExtractor`, `ImageCleaner`, ...).

But for the configuration system, I like the KM3Pipe approach.

Reply:
>  it's hard to then run them in a more simple way (simple for-loop) and debug them, but a lot of it is very nice.

What Tamas showed me on his laptop, it's very easy to debug, because he has stuff like drop into ipython at any point for debugging, if I remember correctly